vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/crawler/transformer/AbstractFetchingUrlsTableTransformer.java

```diff
@@ -66,35 +68,55 @@ protected Set<String> filterResults(Stream<String> urls)
         Stream<String> results = urls;
         if (filterRedirects)
         {
-            Set<String> uniqueUrls = urls.collect(Collectors.toSet());
-            results = uniqueUrls.stream().filter(url -> isNotExistingRedirect(url, uniqueUrls));
+            Set<String> allUrls = urls.collect(Collectors.toSet());
+
+            Set<String> result = new HashSet<>();
+            List<String> redirectChains = new ArrayList<>();
+
+            for (String url : allUrls)
+            {
+                List<String> redirects = getRedirects(url);
+                if (redirects.isEmpty())
+                {
+                    result.add(url);
+                    continue;
+                }
+
+                String redirect = redirects.get(redirects.size() - 1);
+                if (!allUrls.contains(redirect))
+                {
+                    result.add(url);
+                }
+
+                redirectChains.add(url + REDIRECT_JOINER + String.join(REDIRECT_JOINER, redirects));
```

Comment:
the string concatenation should happen at the logging stage, collect all the data into some data structure, e.g. Map